### PR TITLE
DM-29561: Initial version of service-availability-service technote

### PR DESCRIPTION
**Pull request by gpdf**

### CODE REVIEW

File: index.rst [hunk 1]
```diff
@@ -1,26 +1,363 @@
...
+
+We propose the creation of a central availability service, deployed in Kubernetes
+for each RSP instance, and responsible only for responding to VOSI-Availability
+REST queries for all the relevant RSP services on that instance.
+This should of course include every formally IVOA-compliant service, such as TAP
+and SODA, but could usefully be extended to all public-facing RSP services, even
+ones for which VOSI is otherwise irrelevant, e.g., JupyterLab and Firefly.
+
+This service would have a REST interface along the lines of
+``https://(address)/api/central-availability/(service-name)`` .
```
Comment (rra) [inline index.rst]: It looks like we can register these endpoints in the IVOA metadata however we want, so I would be inclined to call this `/status` instead of `/central-availability`, since that's what the rest of the world calls this sort of service so far as I can tell (and it's shorter).

This is an aside, but is the intent to gather all API endpoints for the Science Platform under `/api`? If so, I should change Gafaelfawr, because I was assuming the opposite model (`/service/api` instead of `/api/service`).

File: index.rst [hunk 2]
```diff
@@ -1,26 +1,363 @@
...
+This would allow, for instance, the TAP service to check that the underlying
+Qserv database was responding to simple requests, or the SODA service to check
+that the production Butler is functional.
 
-  Use the following syntax for sections:
+Kubernetes ingress rules would be used to redirect IVOA-style ``/availability``
+endpoints "below" the specific service address,
+such as https://(RSP-external-address)/api/tap/availability,
+away from the service-specific availability endpoint, and
+to the corresponding endpoint on the central availability service.
```
Comment (rra) [inline index.rst:173]: Do we need to do this? The impression I got from the IVOA spec is that everyone is supposed to use the URL in the registry, so we can just point the registry at `/api/status/tap` or whatever.

File: index.rst [hunk 3]
```diff
@@ -1,26 +1,363 @@
...
-  Subsubsections
-  ^^^^^^^^^^^^^^
+Basic /availability XML retrieval from Kubernetes controller
+------------------------------------------------------------
+
+The core element of the proposal is a service which can be configured with
+a list of services to represent, and the logic necessary to query the
+Kubernetes controller for the present state of a service and translate that
+into the ``available`` and ``upSince`` attributes of the VOSI-Availability
+data model.
```
Comment (rra) [inline index.rst]: We lose `upSince` by deferring this Kubernetes work to a later enhancement, but I'm not sure it's that important (it sounds like it's not that frequently used by IVOA services today and thus clients probably don't expect it).

File: index.rst [hunk 4]
```diff
@@ -1,26 +1,363 @@
...
+(very simple) XML format.
+
+Since the author is aware of the existence of anti-XML attitudes in the RSP
+developer community, we note that the central availability service could be
+designed to respond both in JSON and XML.
+One could imagine 
+``https://(address)/api/central-availability/(service-name)``
+responding in JSON and
+``https://(address)/api/central-availability/(service-name)/xml``
+in XML, with the per-service availability endpoints redirected to the latter.
```
Comment (rra) [inline index.rst]: The normal way to do this is with an `Accept` header in the request that specifies either `application/json` or `application/xml`.

File: index.rst [hunk 5]
```diff
@@ -1,26 +1,363 @@
...
+------------------
+
+The VOSI standard states "the availability binding[s] must be available to
+anonymous requests".
+
+In the "naive" implementation model where each service provides its own externally
+accessible availability endpoint, this places each service in the position of 
+handling external data from unauthenticated clients directly.
+This increases the attack surface of the RSP and might require more careful
+service-by-service vetting of their response to availability queries.
```
Comment (rra) [inline index.rst]: While this is true, the return value of the `/availability` endpoint doesn't depend on anything in the query, so the attack surface is limited to fundamental HTTP parsing problems. I think we have as a goal possibly supporting read-only public queries for some services in the future, which would be much more of a risk than the availability endpoint, so I think we're living with this risk anyway.

In other words, I think the merits of the central service are primarily in the roll-up, status UI, and support for non-IVOA services, and the security benefits are fairly minor.

**Review by rra — COMMENTED** [review]
Can we split this into a short-term plan for exposing the IVOA `/availability` endpoints directly (I think that's what we're already doing for TAP) and a longer-term plan that requires the central service? I'd rather hold off on committing to writing the central availability service until we have the resources available for that, and I'm not sure when that would be.

I'm not sure how much value there is in adding the Kubernetes query versus directly querying the `/availability` endpoint of the service. If the pod isn't running, that query will fail anyway, so checking Kubernetes status is both insufficient to know if the service is running and redundant with using the `/availability` endpoint. I think the main value for the Kubernetes-aware check is for reporting status of services with no exposed API. Maybe call that out explicitly and move that part of the design to later work for supporting non-IVOA services?

In other words, I think the implementation strategy here is:

1. Expose the existing `/availability` endpoints for IVOA services.
2. Decide how we want to do health checking for non-IVOA parts of the platform (Gafaelfawr, which is necessary to have correct status for IVOA services, and the Notebook and Portal aspects), and use that to build a status API and UI for the Science Platform more generally
3. Add Kubernetes queries to collect status on components without a status API

**Comment by gpdf** [comment]
Thank you for the detailed review, @rra!  I'll respond in a few spots.

File: index.rst (hunk 5 above)
Comment (gpdf) [reply index.rst]: That makes sense to me.  I'd willing to take out this paragraph altogether if that's what you advise.

File: index.rst (hunk 4 above)
Comment (gpdf) [reply index.rst]: Presumably this would still permit XML to be the default if no `Accept:` header is provided?  I don't _think_ that the IVOA standards require the provision of `Accept:` in order to get conformant responses.

IVOA data-retrieval services (which `/availability` isn't, really) use `&RESPONSEFORMAT=(mime-type-or-other-name)` to ask for specific formats in responses, but that's not an established option for VOSI-availability, so I like your suggestion and I'll reword the proposal around that.

File: index.rst (hunk 3 above)
Comment (gpdf) [reply index.rst]: From a "what's the absolute minimum" perspective, that is surely true, but I view Rubin/LSST as a chance to try to deliver a slightly higher level of service where it seems useful, and if it's not costly in code.  I'll reword the sentence to avoid any implication that the `upSince`-handling is itself a "core" capability.  I was using "core" to mean that the "talk-to-Kubernetes" component is the core of the proposal.

File: index.rst (hunk 2 above)
Comment (gpdf) [reply index.rst:173]: While that is true from a standards-text perspective, the pragmatic reality is that there are quick-and-dirty clients out there - as well as people who just hand-edit URLs - that assume that VOSI-availability is likely to be at an `/availability` endpoint, and I'd like to support that presumption unless we have a positive reason not to.  It's my understanding that adding this rule is a pretty low-cost action for us.

File: index.rst (hunk 1 above)
Comment (gpdf) [reply index.rst]: I'm happy with "status".  When I wrote it, I had in mind that "central-availability" was a meta-name standing in for whatever SQuaRE would want to call it; I'd meant to write something like "(central-availability-name)".  :)

The pathname scheme for the RSP was intended to reflect the three-Aspect model, hence `/portal/...`, `/nb/...`, and `/api/...` - and indeed the TAP service is at `/api/tap` currently.  

The scheme envisioned that all well-advertised public data access Web APIs of the RSP would be under `/api`, but, confusingly, of course, "API" has a more generic meaning than just "member of the API Aspect", and, for instance, REST APIs associated with Firefly or JupyterLab were not intended to have to be under `/api`.

The original scheme also did not address the question of where IAM-related endpoints would appear.  I don't think they need to be under `/api` and maybe even shouldn't be.  Again, the "API Aspect" refers specifically to the user-facing data access interfaces: TAP, SIA, SODA, DataLink, VOSpace, WebDAV, etc.

With that in mind, I wonder whether maybe `/status` itself might be better *not* appearing under `/api`, regardless of what I wrote in line 143.  Perhaps we should talk about this more.  I'd like to understand what the implications for Gafaelfawr actually are.

I see that the relevant technote on endpoints was never published, by the way, my bad.  I'll take a look at getting a review started on that and maybe we can move some of this conversation there.

File: index.rst (hunk 4 above)
Comment (gpdf) [reply index.rst]: Edited accordingly.

File: index.rst (hunk 3 above)
Comment (gpdf) [reply index.rst]: Changed accordingly.

**Comment by gpdf** [comment]
Regarding @rra's comments:

I think Russ' "step 1" is really "step 0": for those services that have a VOSI-availability endpoint (i.e., for TAP), we're already exposing it.  So I don't think it's really part of the scope of this proposal.  However, I'll state more clearly in the note that we currently do this and will continue to do this until a central service is available.

Taking a user's perspective, which is sort of my hat in this enterprise, I don't think getting some sort of 4xx error code from a VOSI-availability call is equivalent to getting a 200 with a suitably detailed "not available" message from the availability service.  The UX is quite different: in the first case, the user has no idea what's wrong and is likely to either a) get frustrated or b) submit a helpdesk ticket saying "your XYZ service is down".  If, as I hope(!), our planned downtime is more than our unplanned downtime, in the central-availability case more often than not the "not available" message will have a meaningful `note` about the downtime, and even a `backAt` for when we expect to be back on the air.

In the Portal, we can immediately make use of that (admittedly, we don't do so now, but that's the effect of the development freeze) to display a corresponding message instead of "failed for unknown reason".  Python interfaces could do the same thing in the long run, as could third-party clients.

This is the same spirit in which we are working on a landing page for the RSP, and framing material for member Aspects' Web interfaces, that can display service alerts and MOTDs.  I think this will have the net effect of improving the UX and of avoiding unnecessary helpdesk tickets, which is a significant contribution to long-term operations scalability.

Regarding prioritization of a rudimentary central-availability service, another reason for it is that, as we deploy additional services of our own devising, we would be able to deploy an MVP that relies on the central service, and not have to implement a service-specific VOSI-availability in a v1.0 of the service.

Just as a matter of philosophy: I think the point of having VOSI-availability at all is not to give the users an (impossible) guarantee "if you submit a query now, it will succeed" (it's impossible for so many reasons: it's not atomic with their actual query, it can't test all possible things the user might do, etc.) but rather to give them, most of the time, when there is a known problem, somewhat more information than just an opaque failure.

File: index.rst (hunk 5 above)
Comment (gpdf) [reply index.rst]: Removed

**Comment by gpdf** [comment]
I've pushed further edits based on @rra's comments, and additional thought.

**Review by rra — APPROVED** [review]
This looks good to me, and thank you for the discussion and clarification. Publish away!
